presto-main/src/main/java/com/facebook/presto/memory/MemoryPool.java

```diff
@@ -130,13 +145,19 @@ public void removeListener(MemoryPoolListener listener)
             }
         }
 
-        onMemoryReserved();
+        onMemoryReserved(queryId);
         return result;
     }
 
-    private void onMemoryReserved()
+    private void onMemoryReserved(QueryId queryId)
     {
-        listeners.forEach(listener -> listener.onMemoryReserved(this));
+        long totalMemoryReservation = queryMemoryReservations.getOrDefault(queryId, 0L) + queryMemoryRevocableReservations.getOrDefault(queryId, 0L);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **154**. At this line the following field is not protected by synchronization: `queryMemoryRevocableReservations`. Your code typically protects this field with synchronization on object: `MemoryPool.this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `MemoryPool.this`. For example, like at lines: **316**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **170**, **247**, **252**, **255**, and **316**.